### DISCUSSION TITLE
Proper way of setting up extension that uses other Quarkus dependencies

### DISCUSSION THREAD

**Discussion by ThomasKluiters** [quarkusio/quarkus#40847] [Q&A]
Hey everyone,

I've been using Quarkus for a good while now. Super happy with it and absolutely LOVE it!

I'm trying to develop my own quarkus extension, though, I'm running into some issues - structuring my project seems quite difficult.

I want to use the following quarkus plugin: https://github.com/quarkiverse/quarkus-openapi-generator to generate some Api clients for my extension. However, this requires me to edit my plugins in the runtime module, and I don't think this is the intention. Is there any suggested approach to set up an extension that leveraging code generators?

**Comment by gsmet** [comment]
Hey @ThomasKluiters,

Glad to hear you love Quarkus. Sorry, we are a bit slow at the moment as a lot of us are in Slovenia for JCon.

So I don't entirely know how the Quarkus OpenAPI Generator actually works but here are a few hints on how we integrate extensions with each other:

- the first important element is the `BuildItem`. They are what we use to pass build information around to `@BuildStep` and they are used to communicate between extensions. A lot of extensions provide build items that can be used to integrate with them. (I don't know if this particular extension provides integration points - but see below)
- another point is that you don't always need to do anything. If your extension for instance contains some classes that should be automatically handled by the other extension if they were in a regular app, the only thing probably missing is that you need to push the classes to the Jandex index (you have a build item for that (`AdditionalIndexedClassesBuildItem`) if you want to target a few specific classes or you can also use the Jandex Maven plugin to index everything).
- I don't know anything about your extension so mostly handwaving here, but if you are generating classes, you will also have to push them to the index of you want them to be taken into account.

So in short: maybe there's nothing to do except push the classes to the Jandex index, maybe there are integration points that you can use in the extension and maybe you will need more but, in this case, better ping the OpenAPI generator team in an issue (and give a lot more details so that they can figure out what you need).

HTH!
